zulip_bots/zulip_bots/bots/connect_four/test_connect_four.py

```diff
@@ -0,0 +1,483 @@
+        start_one_player_data = {'state': 'playing', 'game_type': 'one_player', 'board': blank_board, 'players': ['foo@example.com', 'The Computer'], 'turn': 0}
+
+        # Test the static messages
+        self.verify_dialog([('help', help_message())])
+        self.verify_dialog([('status', no_game_status())])
+
+        # Test 'start game with' command
+        verify_response('start game with', 'Sorry, I couldn\'t determine who you wanted to start the game with.\n'
+                        'Type ```start game with user@example.com``` to start a game with another user,\n'
+                        'or type ```start game with computer``` to start a game with the computer', 0)
+        verify_response('start game with @**foo**', 'Sorry, I couldn\'t determine the user you'
+                        'specified. Make sure you specify the user as\n'
+                        '```user@example.com```.', 0)
+        verify_response('start game with computer', '**You started a new game with the computer!**', 0)
+        verify_response('start game with user@example.com', 'Inviting user@example.com to play Connect Four with you\n'
+                        'I\'ll let you know when they respond to the invitation', 0)
+
+        # Test 'move' commands
+        verify_response('move', 'Sorry, I couldn\'t determine the move you wanted to make\n'
+                        'Type ```move <column-number>``` to specify your move', 0, data=start_two_player_data)
+        verify_response('move 8', 'That\'s an invalid move. Please specify a column' +
+                        ' with at least one blank space, between 1 and 7', 0, data=start_two_player_data)
+        verify_response('move 1', 'You placed your token in column 1.', 0, data=start_two_player_data)
+        verify_response('move 1', '**The computer moved in column 1**', 3, data=start_one_player_data, computer_move=0)
+
+        # Test game over phase
+        verify_response('move 1', '**Congratulations, you win!**', 4, data=end_two_player_data)
+
+        # Test 'quit' command
+        verify_response('quit', 'Are you sure you want to quit? You will forfeit the game!\n'
+                        'Type ```confirm quit``` to forfeit.', 0, data=start_two_player_data)
+        verify_response('quit', 'You\'re invitation for foo2@example.com'
+                        ' is still pending!\nTo quit that game, type ```cancel game```', 0, data=inviting_two_player_data)
+
+        # Test 'confirm quit' command
+        verify_response('confirm quit', '**You have forfeit the game**\nSorry, but you lost', 0, data=start_two_player_data)
+        verify_response('confirm quit', 'It doesn\'t seem there is game to quit...', 0, data=inviting_two_player_data)
+
+        # Test 'lack of privelege'
+        verify_response('move', 'Sorry, but you don\'t have privilege to run that command.'
+                        'Type ```status``` to see the current status of the bot.', 0, data=inviting_two_player_data)
+
```

Comment:
@sColin16: These tests look great! I have a couple of concerns:

* I think the helper functions such as `verify_response`, `help_message` should be outside the `test_bot` message.
* Also, instead of having one large `test_bot` function that tests everything, I would recommend splitting that into many test functions, perhaps one for each command? That would just be easier to read and more organised! :)